modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,173 @@
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, NotTipAcceptedBlockNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def updateTipParentsToRemove(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
+                (accepted.addUsage(hashedBlock).unsetAsTip.some, ().asRight)
+              case unexpected =>
+                (unexpected, TipBlockToUnsetAsTipNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def updateTipParentsToRemain(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, true)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, TipBlockToRemainAsTipNotFound(hash, unexpected).asLeft)
+            }
+            .flatMap(_.liftTo[F])
+        }.void
+
+      private def tryAddAsATip(condition: Boolean, hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        if (condition)
+          blocks(hashedBlock.proofsHash).modify {
+            case Some(accepted @ AcceptedBlock(_, _, usages, false))
+                if usages.size <= tipsConfig.maximumTipUsages - 1 =>
+              (accepted.setAsTip.some, ())
+            case other => (other, ())
+          } else Sync[F].unit
+    }
+
+  sealed trait AcceptedBlockState
+  case object Accepted extends AcceptedBlockState
+  case object InMajority extends AcceptedBlockState
+  sealed trait StoredBlock
+  case class WaitingBlock(block: Signed[DAGBlock], usages: Set[ProofsHash]) extends StoredBlock
+  case class AcceptedBlock(block: Hashed[DAGBlock], state: AcceptedBlockState, usages: Set[ProofsHash], isTip: Boolean)
+      extends StoredBlock {
+    def addUsage(hashedBlock: Hashed[DAGBlock]): AcceptedBlock = this.focus(_.usages).modify(_ + hashedBlock.proofsHash)
+    def setAsTip: AcceptedBlock = this.focus(_.isTip).replace(true)
+    def unsetAsTip: AcceptedBlock = this.focus(_.isTip).replace(false)
+  }
+
+  sealed trait BlockServiceError extends NoStackTrace
+  sealed trait TipsUpdateError extends BlockServiceError
+  case class TipBlockToUnsetAsTipNotFound(hash: ProofsHash, got: Option[StoredBlock]) extends TipsUpdateError {
+    override def getMessage: String = s"Tip block to unset as tip with hash=$hash not found! But got: $got"
```

Comment:
why not just `Tip not found` ?

Reply:
Because I would like to have a concrete error for every case. If I have more generic `TipNotFound` I won't be able to tell during which update functionality (remaining or removing as a tip) it happened. I think we are allowed to have such precise error messages in the implementation.